### ISSUE TITLE
Promtail/grafana-agent on k8s splits and applies incorrect labels on long log messages

### ISSUE THREAD

**Issue by martinjaegerdk** [grafana/loki#8112]
**Describe the bug**
Long log lines produced by a pod in k8s, sometimes get's broken into smaller pieces, and some pieces show up in Loki with wrong labels.

1. `POD-A` in `NAMESPACE-A` on `NODE-A` emits a long log message (>20.000 characters)
2. In Loki, that log message sometimes shows up on other pods in other namespaces but still on `NODE-A`, other times it shows up on `POD-A`. Sometimes it’s only part of the log message that appears in one `POD`, and the rest in on another `POD`.

**To Reproduce**
To test this out, i have deployed a `POD` which emits a long logline in this format: `longlogstart/{id}-{{randChars(2000)}}-longlogend/{id}`. This allows me to see when the log should start and end, and also if the logline get’s broken up and is shown with different labels. (the `id` helps with this correlation)

**Go code used**
```golang
var letters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")

func randSeq(n int) string {
	b := make([]rune, n)
	for i := range b {
		b[i] = letters[rand.Intn(len(letters))]
	}
	return string(b)
}

logid := randSeq(10)
log.Printf("longlogstart/" + logid + "-" + randSeq(20000) + "longlogend/" + logid)
```

**Observations**
- The log message (or just parts of it) always appears on the same `NODE` where the originating pod is running (which makes me suspect the agent, and not Loki backend)
- The logline only appears in it’s entirety on the originating pod (including start and end). Meaning: `{k8s_cluster=“c1”, pod!=“POD-A”} |= “longlogstart” |=“longlogend”` is empty
- When the logline appears only including `longlogstart` (not `longlogend`), there is always a real log message at the end of it (from the correct pod). Meaning that it mixes it in. The first part of that logline (the long one) is always cut of after **16,364** chars. After that comes the real log on a new line. 
- When the logline appears only including `longlogend` it’s always on the originating pod (the correct one). Meaning: `{k8s_cluster=“c1”, pod!=“POD-A”} |= “longlogend” !=“longlogstart”` is empty

**Expected behavior**
Logs from different pods should not be mixed

**Environment:**
 - Infrastructure: k8s (running on Azure Kubernetes Service)

**Screenshots, Promtail config, or terminal output**
We run grafana-agent (`grafana/agent:v0.29.0`) in a DaemonSet, using below `agent.yaml`

```yaml
server:
  log_level: info

logs:
  configs:
    - clients:
        - url: http://redacted/loki/api/v1/push
          oauth2:
            client_id: $${OAUTH2_CLIENT_ID}
            client_secret: $${OAUTH2_CLIENT_SECRET}
            token_url: redacted
            scopes:
              - "redacted"
          external_labels:
            k8s_cluster: ${CLUSTER_NAME}
            region: ${REGION_NAME}
            env: ${ENVIRONMENT_NAME}
            product_id: redacted
            provider: ${PROVIDER_NAME}
      name: default
      scrape_configs:
        - job_name: kubernetes-pods
          kubernetes_sd_configs:
            - role: pod
          pipeline_stages:
            - cri: {}
          relabel_configs:
            - source_labels:
                - __meta_kubernetes_pod_controller_name
              regex: ([0-9a-z-.]+?)(-[0-9a-f]{8,10})?
              action: replace
              target_label: __tmp_controller_name
            - source_labels:
                - __meta_kubernetes_pod_label_app_kubernetes_io_name
                - __meta_kubernetes_pod_label_app
                - __tmp_controller_name
                - __meta_kubernetes_pod_name
              regex: ^;*([^;]+)(;.*)?$
              action: replace
              target_label: app
            - source_labels:
                - __meta_kubernetes_pod_label_app_kubernetes_io_component
                - __meta_kubernetes_pod_label_component
              regex: ^;*([^;]+)(;.*)?$
              action: replace
              target_label: component
            - action: replace
              source_labels:
                - __meta_kubernetes_pod_node_name
              target_label: node_name
            - action: replace
              source_labels:
                - __meta_kubernetes_namespace
              target_label: namespace
            - action: replace
              replacement: $1
              separator: /
              source_labels:
                - namespace
                - app
              target_label: job
            - action: replace
              source_labels:
                - __meta_kubernetes_pod_name
              target_label: pod
            - action: replace
              source_labels:
                - __meta_kubernetes_pod_container_name
              target_label: container
            - action: replace
              replacement: /var/log/pods/*$1/*.log
              separator: /
              source_labels:
                - __meta_kubernetes_pod_uid
                - __meta_kubernetes_pod_container_name
              target_label: __path__
            - action: replace
              regex: true/(.*)
              replacement: /var/log/pods/*$1/*.log
              separator: /
              source_labels:
                - __meta_kubernetes_pod_annotationpresent_kubernetes_io_config_hash
                - __meta_kubernetes_pod_annotation_kubernetes_io_config_hash
                - __meta_kubernetes_pod_container_name
              target_label: __path__
            - action: replace
              source_labels:
                - __meta_kubernetes_pod_label_product_id
                - __meta_kubernetes_pod_label_product
              regex: ^;*([^;]+)(;.*)?$
              target_label: product_id
            - action: replace
              source_labels:
                - __meta_kubernetes_pod_label_env
                - __meta_kubernetes_pod_label_environment
              regex: ^;*([^;]+)(;.*)?$
              target_label: env
  positions_directory: /tmp/positions

```

**Comment by JohanLindvall** [comment]
FWIW, we have seen this bug with promtail 2.7.0. We worked around it by switching to vector as an ingester.

**Comment by martinjaegerdk** [comment]
> FWIW, we have seen this bug with promtail 2.7.0. We worked around it by switching to vector as an ingester.

Thanks for that. I had no idea that existed :) It's useful to hear that it's not only a thing we see.

**Comment by gasmick** [comment]
unfortunately happens with Promtail versions 2.7.1 and 2.7.3

**Comment by dbluxo** [comment]
@martinjaegerdk We have also observed the problem since Promtail 2.7.x. Before giving more details, I would be interested to know which [container runtime](https://kubernetes.io/docs/setup/production-environment/container-runtimes/) you are using? CRI-O?

**Comment by martinjaegerdk** [comment]
@dbluxo 
```
OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
Ubuntu 18.04.6 LTS   5.4.0-1094-azure   containerd://1.6.4+azure-4
```

**Comment by dbluxo** [comment]
We can observe the problem when we update Promtail from 2.6.1 to 2.7.x that parts of log messages end up in the logs of other pods. We are currently investigating [this](https://github.com/grafana/loki/pull/6177) PR as a possible cause, but this would only affect the CRI-O container runtime we are running on.

**Comment by dbluxo** [comment]
Ah, it is not important on which container runtime you are running on. Have you also configured the [cri stage](https://grafana.com/docs/loki/latest/clients/promtail/stages/cri/)? Then [this](https://github.com/grafana/loki/pull/6177) PR would also apply.

**Comment by dbluxo** [comment]
Update: The problem was definitely introduced with [this](https://github.com/grafana/loki/pull/6177) PR.
Namely, it is not ensured that before being merged ([here](https://github.com/grafana/loki/blob/v2.7.3/clients/pkg/logentry/stages/extensions.go#L65) or [here](https://github.com/grafana/loki/blob/v2.7.3/clients/pkg/logentry/stages/extensions.go#L76)) the partial log lines, which end up in [this](https://github.com/grafana/loki/blob/v2.7.3/clients/pkg/logentry/stages/extensions.go#L56-L83) routine, are labeled identically, i.e. have the same label values (filename, app, instance, container, pod etc.).

**Comment by martinjaegerdk** [comment]
> Have you also configured the [cri stage](https://grafana.com/docs/loki/latest/clients/promtail/stages/cri/)? Then [this](https://github.com/grafana/loki/pull/6177) PR would also apply.

Yes. We have configured the cri stage, like so:
```yaml
    pipeline_stages:
      - cri: {}
```
@dbluxo thanks for looking into this

**Comment by vanugrah** [comment]
Hey folks, we're running loki and promtail at significant scale and I can also confirm that upgrading from promtail 2.6.1 to 2.7.1 lead to log streams from different workloads to be intermingled. We're also using the CRI-O container runtime. I'm not sure if this is important to note or not but we observed that this log "leak" was bidirectional (logs from pod A would end up in the stream for pod B and vice-versa). We observed two separate instances of this occurring. 

After reverting promtail back to 2.6.1 we are no longer seeing this behavior.
